### PR TITLE
DistanceMap rework (re-PR of #6940)

### PR DESCRIPTION
**Pull request by warp-core**
**Bugfix:**

This PR takes @alextd's branch from #6940, and brings it up to date and addresses some of the comments on that PR.

## Fix Details
In addition to the changes from #6940, I have moved the new `RoutePlan` class out of the DistanceMap files into its own files.
I have also then moved `DistanceMap::Edge` to its own files and renamed it `RouteEdge` because `RoutePlan.h` needs the class to be fully defined but including all of `DistanceMap.h` there doesn't seem great.

## Testing Done
None beyond what was done for the previous PR, and given all the conflicts and my changes, this probably needs more testing, or at least, more reviewing.

## Save File
This save file can be used to verify the bugfix. The bug will occur when using {{insert commit hash / version}}, and will not occur when using this branch's build.
N/A? Just play the game.

### CODE REVIEW

**Comment by warp-core** [comment]
@quyykk are you killing the CodeBlocks project or do I gotta fix this?

File: source/RouteEdge.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+	// There could be a System *thisSystem, but it would remained unused.
+	const System *prev = nullptr;
+	// Fuel/days needed to get to this system using the route through 'prev'.
+	int fuel = 0;
+	int days = 0;
+	// Danger tracks up to the 'prev' system, not to the this system.
+	// It's used for comparison purposes only. Anyone going to this system
+	// is going to hit its danger anyway, so it doesn't change anything.
+	double danger = 0.;
```
Comment (RisingLeaf) [inline source/RouteEdge.h:47]: shouldnt this be private?

File: source/RouteEdge.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "RouteEdge.h"
+
```
Comment (RisingLeaf) [inline source/RouteEdge.cpp:17]: so, according to the c++ style guide we use:

> The std namespace
> ▶ Each .cpp file should put "using namespace std;" immediately after the #includes.

Although I am not sure if it makes sense if we do not include from std

Comment (warp-core) [reply source/RouteEdge.cpp:17]: I don't think it does make sense to use it if no members of `std` are actually used.

File: source/RouteEdge.h (hunk 1 above)
Comment (alextd) [reply source/RouteEdge.h:47]: I'd say RouteEdge should actually just be a struct

**Comment by alextd** [comment]
Hiho hello, I am back and have re-merged this into my original branch @ https://github.com/alextd/endless-sky/tree/overhaul-distance-map. A few more merge conflicts handled.

(Though noticeably the main reason for this branch's changes, the "jump range" value has been removed, so that could certainly be re-introduced with this merged in. Also the secondary fix with escorts refueling was addressed in another simpler update sometime last year.)

**Comment by Ferociousfeind** [comment]
It should probably be noted that there are systems in vanilla ES (the Umbral Reach) which escorts could get lost in, and this is practically a requirement for the Avgi (PR #9172) since their space is (almost) completely one-way connected to the rest of the galaxy- once you jump in, you cannot jump out, which means that escorts cannot jump out _OR_ in.

I think there's one or two bidirectional lanes over on the ember waste side of avgi space, or other ways to force my escorts to follow me into avgi space, but it's a whole hassle. (Not even commanding my fleet to make a single jump with me will force them to follow me to avgi space)

**Comment by TomGoodIdea** [comment]
There are merge conflicts ~~hmm maybe because the last sync with master was more than a year ago~~

**Comment by alextd** [comment]
Taking a look tonight...

**Comment by alextd** [comment]
Took a nice long delay but I got this up to latest master.

I think I'll reopen the original PR using my branch: #6940

**Pull request closed without merging**